### DISCUSSION TITLE
Improving the UX of dagger input

### DISCUSSION THREAD

**Discussion by aluzzardi** [dagger/dagger#525] [General]
I find the `dagger input` commands a bit cumbersome, there's a lot to type and somehow I always type the wrong thing.

What are your thoughts about using flags rather than sub-commands to set the different input types? The obvious downside is "flag mess", but I don't know if that's worse than what we have now.

Examples:

Regular text input (default)
```
dagger set foo bar
```

git
```
dagger set repository --git https://github.com/dagger/dagger.git --git-ref main

OR

dagger set repository --git 'https://github.com/dagger/dagger.git#main'
```

dir
```
dagger set foo --dir . --exclude node_modules
```

secret
```
dagger set --secret foo bar
```

list inputs (e.g. `dagger input list`):
```
# `dagger set` with no arguments will print all available inputs
dagger set 
```

interactive mode (e.g. `dagger input edit`):
```
dagger set -i
```

Other things:
- `dagger input unset` would just become `dagger unset`
- `dagger query` could perhaps become `dagger get`? Or, for symmetrical reasons, `dagger get` could retrieve inputs (e.g. `dagger set foo bar && dagger get foo`)

**Comment by samalba** [comment]
👍 on this, it's also aligned with what @shykes suggested earlier.

We could also auto-detect the value (string, json, directory) to omit the flag or raise a warning if the flag is passed.

**Comment by shykes** [comment]
I agree that the current UI for inputs needs improvement.

A few questions and comments:

> *`dagger set` instead of `dagger input`

OK with me.

> *`dagger input unset` would just become `dagger unset`*

Makes perfect sense.

> *[Use flags instead of sub-commands]*

I don’t hate it, but it doesn’t solve the “lots to type” problem. All commands are the same length or longer, except for the special case `dagger set foo bar`. It feels more like a subjective preference to switch to flags.

> `dagger set repository --git https://github.com/dagger/dagger.git --git-ref main`

In this example, is `—git` a boolean flag, or a string flag?

> *interactive mode (e.g. dagger input edit):*

Do we still want to implement an interactive command-line wizard? If so, “edit” and “interactive” would be two different modes, right?

> *# `dagger set` with no arguments will print all available inputs*

I don’t find that intuitive. I would actually expect `dagger set` to open the editor or interactive wizard, at least when in a terminal.

> *`dagger query` could perhaps become `dagger get`?*

That’s a whole other can of worms… IMO better if we discuss that one separately.
